packages/expo-application/android/src/main/java/expo/modules/application/ApplicationModule.java

```diff
@@ -0,0 +1,159 @@
+package expo.modules.application;
+
+import java.text.SimpleDateFormat;
+import java.util.Calendar;
+import java.util.HashMap;
+import java.util.Map;
+
+import android.app.Activity;
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.content.pm.PackageInfo;
+import android.content.pm.PackageManager;
+import android.os.RemoteException;
+import android.provider.Settings;
+import android.util.Log;
+
+import com.android.installreferrer.api.InstallReferrerClient;
+import com.android.installreferrer.api.InstallReferrerStateListener;
+import com.android.installreferrer.api.ReferrerDetails;
+
+import org.unimodules.core.ExportedModule;
+import org.unimodules.core.ModuleRegistry;
+import org.unimodules.core.Promise;
+import org.unimodules.core.interfaces.ActivityProvider;
+import org.unimodules.core.interfaces.ExpoMethod;
+import org.unimodules.core.interfaces.RegistryLifecycleListener;
+
+public class ApplicationModule extends ExportedModule implements RegistryLifecycleListener {
+  private static final String NAME = "ExpoApplication";
+  private static final String TAG = ApplicationModule.class.getSimpleName();
+
+  private ModuleRegistry mModuleRegistry;
+  private Context mContext;
+  private ActivityProvider mActivityProvider;
+  private Activity mActivity;
+
+  public ApplicationModule(Context context) {
+    super(context);
+    mContext = context;
+  }
+
+  @Override
+  public String getName() {
+    return NAME;
+  }
+
+  @Override
+  public void onCreate(ModuleRegistry moduleRegistry) {
+    mModuleRegistry = moduleRegistry;
+    mActivityProvider = moduleRegistry.getModule(ActivityProvider.class);
+    mActivity = mActivityProvider.getCurrentActivity();
+  }
+
+  @Override
+  public Map<String, Object> getConstants() {
+    HashMap<String, Object> constants = new HashMap<>();
+
+
+    String applicationName = mContext.getApplicationInfo().loadLabel(mContext.getPackageManager()).toString();
+    String packageName = mContext.getPackageName();
+
+    constants.put("applicationName", applicationName);
+    constants.put("bundleId", packageName);
+
+    PackageManager packageManager = mContext.getPackageManager();
+    try {
+      PackageInfo pInfo = packageManager.getPackageInfo(packageName, 0);
+      constants.put("nativeAppVersion", pInfo.versionName);
+      constants.put("nativeBuildVersion", pInfo.versionCode);
+    } catch (PackageManager.NameNotFoundException e) {
```

Comment:
Under what conditions can this happen? We should document that..

Reply:
This error is thrown "when a given package, application, or component name cannot be found". I don't really expect this to fail since we grab the package name from `mContext`, so we should expect it to be the correct `bundleId` of the package. However, `getPackageInfo()` throws this error, so we have to catch it in a `catch` block.